PWGLF/TableProducer/Strangeness/sigma0builder.cxx

```diff
@@ -756,83 +900,134 @@ struct sigma0builder {
+        histos.fill(HIST("GeneralQA/hCentralityVsInteractionRate"), centrality, interactionRate);
+      }
+
+      std::vector<int> bestGammasArray;
+      std::vector<int> bestLambdasArray;
+
+      //_______________________________________________
+      // Photon-only loop
+      for (auto& gamma : V0s) { // selecting photons from Sigma0
         if (!gamma.has_v0MCCore())
           continue;
 
         auto gammaMC = gamma.v0MCCore_as<soa::Join<aod::V0MCCores, aod::V0MCCollRefs>>();
 
-        // Auxiliary histograms:
         if (gammaMC.pdgCode() == 22) {
-          histos.fill(HIST("MC/h3dGammasXYZ"), gamma.x(), gamma.y(), gamma.z());
+          histos.fill(HIST("MC/h2dGammaXYConversion"), gamma.x(), gamma.y());
           float GammaY = TMath::Abs(RecoDecay::y(std::array{gamma.px(), gamma.py(), gamma.pz()}, o2::constants::physics::MassGamma));
-          float gammaMCpT = RecoDecay::pt(array{gammaMC.pxMC(), gammaMC.pyMC()});
           if (GammaY < 0.5) {                                                                                                                // rapidity selection
-            histos.fill(HIST("MC/h2dPtVsCentrality_GammaBeforeSel"), centrality, gamma.pt());                                                // isgamma
-            if (gammaMCpT > 0)
-              histos.fill(HIST("MC/h2dGammaPtResolution"), 1.f / gammaMCpT, gamma.pt() - gammaMCpT); // pT resolution
-
-            if (gammaMC.pdgCodeMother() == 3212) {
-              histos.fill(HIST("MC/h2dPtVsCentrality_GammaSigma0"), centrality, gamma.pt()); // isgamma from sigma
-            }
-            if (gammaMC.pdgCodeMother() == -3212) {
-              histos.fill(HIST("MC/h2dPtVsCentrality_GammaAntiSigma0"), centrality, gamma.pt()); // isgamma from sigma
-            }
+            histos.fill(HIST("MC/h2dPtVsCentralityBeforeSel_MCAssocGamma"), centrality, gamma.pt());                                         // isgamma
           }
         }
-        if (gammaMC.pdgCode() == 3122) { // Is Lambda
-          float LambdaY = TMath::Abs(RecoDecay::y(std::array{gamma.px(), gamma.py(), gamma.pz()}, o2::constants::physics::MassLambda));
-          float lambdaMCpT = RecoDecay::pt(array{gammaMC.pxMC(), gammaMC.pyMC()});
-          if (LambdaY < 0.5) { // rapidity selection
-            histos.fill(HIST("MC/h2dPtVsCentrality_LambdaBeforeSel"), centrality, gamma.pt());
-            if (lambdaMCpT > 0)
-              histos.fill(HIST("MC/h2dLambdaPtResolution"), 1.f / lambdaMCpT, gamma.pt() - lambdaMCpT); // pT resolution
-            if (gammaMC.pdgCodeMother() == 3212) {
-              histos.fill(HIST("MC/h2dPtVsCentrality_LambdaSigma0"), centrality, gamma.pt());
-            }
-          }
+
+        // basic photon selection
+        if (!processPhotonCandidate(gamma))
+          continue;
+
+        // Save indices of best gamma candidates
+        bestGammasArray.push_back(gamma.globalIndex());
```

Comment:
Here you have two O(N) loops over V0s - one for lambdas, one for photons. It would be significantly better performance-wise to just do all-at-once (i.e. `if(){populate vector}` inside one single for loop)